### ISSUE TITLE
Refactor initialisation of invariants in value refs to happen after parsing

### ISSUE THREAD

**Issue by agrrr3** [freeorion/freeorion#3225]
Code Improvement request
===============

Idea
----
Parsing is happening in parallel. Initialisation of invariants currently happens at creation time of that value refs. To correctly figure out invariants and related derived values, value refs using NamedLookups need the looked up value ref to exists on creation time. There is a workaround to sleep and retry a few times which is ugly but should work in most cases (#3221).

The clean solution is to initialise the invariants after all parsing is finished.

Gain
----
Deterministic invariant resolution. No unnecessary waiting time. The deterministic nature helps to narrow down scope of bugs like #3220

**Comment by o01eg** [comment]
> The clean solution is to initialise the invariants after all parsing is finished.

I suppose we can use `Pending::WaitForPending` before initialize the invariants.

**Comment by agrrr3** [comment]
> > The clean solution is to initialise the invariants after all parsing is finished.
> 
> I suppose we can use `Pending::WaitForPending` before initialize the invariants.

Not sure how that would work out. Could you sketch an implementation idea?

**Comment by o01eg** [comment]
Yes, in `IApp::StartBackgroundParsing` values ref should be moved last, also each Pending should be saved in variables and each variables should be waited before values ref parsing.

**Comment by agrrr3** [comment]
> Yes, in `IApp::StartBackgroundParsing` values ref should be moved last, also each Pending should be saved in variables and each variables should be waited before values ref parsing.

I guess you mean the parsing of named_values.focs.txt . Why I put this first? Because most everything using NamedLookup depends on that. So rather than putting it last we could delay async parsing of the other files until named_values is finished, but....

...but that is also only a mitigation, you can define named values inline (named-in-the-middle) and use NamedLookup on these. Yet though did not encounter a case where that wasnt a code smell. 

@Oberlus did you need to NameLookup a named-in-the-middle valueref yet?

the definition/parsing of named values (NamedReal,...) is not the problem, the use of it is (NamedRealLookup,...)

probably i still misunderstood your suggestion..


maybe something like this?

* put all pending containers into variables
* then after  after  calling `WaitForPending(named_value_ref_parser_future);`
** WaitForPending for all variables
** then trigger initialisation on all value refs using the variables and pass those to the managers (e.g. `GetTechManager().SetTechs`) - 

delayed initilisation means to trigger initialisation not from value ref constructor but from a member function. and i also need that not only on valuerefs but everything which can contain value refs (conditions, effects).

rather than triggering initialisation directly I guess that should be done in the managers e.g. in CheckPendingTechs as those managers are supposed to know about where in a e.g. Tech object one finds conditions,effectgroups,valuerefs e.g. research_cost, research_turns, unlockable items,effects, ) or it maybe the object themselves should expose that function.

**Comment by Oberlus** [comment]
> did you need to NameLookup a named-in-the-middle valueref yet?

I didn't need it, but I didn't know it was possible.
So, I could do either of the following, right?

* Using named_values.macros:
  - NamedValue XXX in name_values.macros.
  - NamedLookup XXX in THIS_EFFECT.focs.
  - NamedLookup XXX in THAT_EFFECT.focs.

* Not using named_values.macros:
  - NamedValue XXX in THIS_EFFECT.focs.
  - NamedLookup XXX in THAT_EFFECT.focs.

**Comment by agrrr3** [comment]
yes, you could do both ways.

I just dont see a legitimate use case for the second one (not using `named_values.focs.txt`).. 

Maybe for a plugin or if you use the same NamedValue in the same module/file

**Comment by Oberlus** [comment]
> I just don't see a legitimate use case for the second one (not using `named_values.focs.txt`)..

I would use the one that is better in terms of (first) performance and code readability (second).

It seems it would be much easier, in general, to avoid using Lookups when possible (such as what you did for the disconnected influence malus: two different named values referencing same macro constant, nice).
IMPERIAL_GARRISON_MAX_TROOPS_FLAT (used in two FOCS files) could use the same trick. Maybe others.

NUM_COMBAT_ROUNDS/_FIGHTERS are example of named values that must be in named_values.macros (only referenced from stringtables, doesn't appear in FOCS files).

SHP_REINFORCED_HULL_BONUS is only referenced from one FOCS effect, so it could be removed from named_values.macros and just define it in the FOCS file (no Lookups), right?

**Comment by agrrr3** [comment]
> > I just don't see a legitimate use case for the second one (not using `named_values.focs.txt`)..
> 
> I would use the one that is better in terms of (first) performance and code readability (second).

performance is the same in both cases. Under the hood when parsing e.g. a named-in-the-middle NamedReal, the value ref gets registered in the double registry and then a NamedLookup valueref gets inserted in the containing valueref. The same NamedLookup valueref which gets inserted when having NamedRealLookup in FOCS.

So code readability/maintainability. If it is used only once make it inline/named-in-the-middle. If you use at different places in FOCS put it in named_values.focs.txt
> It seems it would be much easier, in general, to avoid using Lookups when possible (such as what you did for the disconnected influence malus: two different named values referencing same macro constant, nice).
> IMPERIAL_GARRISON_MAX_TROOPS_FLAT (used in two FOCS files) could use the same trick. Maybe others.

No, if it is the same value it should be only one NamedReal. We only really need the named values if we want to access it from AI or if it is a calculation result (for the constant value a macro suffices). Although the named values pedia entry gives a nice quantative overview over the effects in the game.

> SHP_REINFORCED_HULL_BONUS is only referenced from one FOCS effect, so it could be removed from named_values.macros and just define it in the FOCS file (no Lookups), right?

Yes, that would be good. I had it there only as a dummy/test entry in the beginning.

**Comment by agrrr3** [comment]
> Yes, in `IApp::StartBackgroundParsing` values ref should be moved last, also each Pending should be saved in variables and each variables should be waited before values ref parsing.

> maybe something like this?
> 
>     * put all pending containers into variables
> 
>     * then after  after  calling `WaitForPending(named_value_ref_parser_future);`
>       ** WaitForPending for all variables
>       ** then trigger initialisation on all value refs using the variables and pass those to the managers (e.g. `GetTechManager().SetTechs`) -

@o01eg i got a complication saving the Pending. The pending need to be in the Managers, as e.g. calculation of checksum needs to wait until parsing finishes using the Pending future. Can I also somehow wait `IApp::StartBackgroundParsing` on that thread using the future? do i need to use shared_future? or should i set a flag in all the managers like `parsing_finished` which i can query from IApp?

Also WaitForPending does not seem to be thread safe. See #3233

**Comment by o01eg** [comment]
@agrrr3 In #3211 I used `std::promise` to use as thread safe barrier to wait in main thread for named value parser finished.

> Can I also somehow wait IApp::StartBackgroundParsing on that thread using the future?

No, parsers create new threads with `std::async`.

**Comment by agrrr3** [comment]
> @agrrr3 In #3211 I used `std::promise` to use as thread safe barrier to wait in main thread for named value parser finished.

@o01eg Oh nice. Didnt see that. Hard to wrap my head around what a `Pending` object with a `Parsing` member exactly does (i think it passes the parsing result through like Pending without a Parsing and it signals after the parse function returned). 

Waiting only for the NamedValueRefParser won't suffice though - you/we need to wait for all the parsers which may parse valuerefs to finish.

> 
> > Can I also somehow wait IApp::StartBackgroundParsing on that thread using the future?

> 
> No, parsers create new threads with `std::async`.

i meant if there is a way to wait in IApp::StartBackgroundParsing on the future contained in the Pending. Actually i think I could by adding some function to Pending which only `wait()` but not `get()`. If that works, I guess it would provide the same functionality like your Pending/Parsing implementation, but with less structural overhead.

**Comment by o01eg** [comment]
> Waiting only for the NamedValueRefParser won't suffice though - you/we need to wait for all the parsers which may parse valuerefs to finish.

I only re-implemented current `WaitForPending` in more multi-thread solution.

**Comment by agrrr3** [comment]
Maybe this(?) or something related caused an unsufferable delay in #3931 when the named values were declared in on field declaration and looked up in another.. but it was working well when using the suggested pattern of declaring such in named_values.focs.txt ( https://github.com/freeorion/freeorion/commit/6e89d42bc2abce2cddfb90c764b873991ac801ca ).

Anyway that (delay or failure to lookup) should not have happened.

**Comment by o01eg** [comment]
Maybe we can remove parallel parsing because we still had to do it for Python parsing.

**Comment by geoffthemedio** [comment]
> Maybe we can remove parallel parsing because we still had to do it for Python parsing.

could you rephrase that or replace "it" with what "it" is?

**Comment by o01eg** [comment]
Python FOCS parser is single-threaded so all parsers are going to be sequential either way.
